### PR TITLE
Account for FOV in fog

### PR DESCRIPTION
**Pull request by rreusser**
This PR adds an adjustment to the fog range so that the appearance is unchanged when the camera FOV changes. The correction comes from this single line:

https://github.com/mapbox/mapbox-gl-js/blob/532e92724a71487e6dab6c6be4a1c824efaf6c9e/src/geo/transform.js#L1287

Since fog range is divided by the window height, that means we only need to shift the range by `0.5 / tan(fov / 2)` as the fov changes. Everything is piped through the same fog helper so that this change applies to symbols, terrain, and culling.

A choice made here:
The camera position moves as the FOV changes, but we want the fog to cover the map identically as the fov changes. A clean choice is fog distance := 0 is the camera center, but this would result in fog start being generally somewhat negative. Instead, this PR anchors zero at the position of the default camera. As you decrease the FOV, the camera moves farther away, but fog distance = 0 stays fixed at the reference (default-fov) camera position and the appearance of the fog is unchanged.

Before:
badfov

After:
fov

### CODE REVIEW

**Comment by rreusser** [comment]
Oh, and I did consider multiplicative scaling so that the camera position would be zero and the camera center would be 1. This seems tempting, but when the fov decreases and the camera backs up, the far limit gets pushed very var away. Fog pushed away when you narrow the field of view is precisely the opposite of what you'd want:

Multiplicative scaling (camera = 0, center = 1):
multiplicative

Additive shifting (shifting to maintain settings relative to default camera):
additive

File: src/style/fog_helpers.js [hunk 1]
```diff
@@ -33,12 +35,21 @@ export function getFogOpacity(state: FogState, pos: Array<number>, pitch: number
     return falloff * fogPitchOpacity;
 }
 
+export function getFovAdjustedFogRange(state: FogState, fov: number): [number, number] {
+    // This function computes a shifted fog range so that the appearance is unchanged
+    // when the fov changes. We define range=0 starting at the camera position given
+    // the default fov. We avoid starting the fog range at the camera center so that
+    // ranges aren't generally negative unless the FOV is modified.
```
Comment (karimnaaji) [inline src/style/fog_helpers.js:42]: cc @ansis about the conversation we just had concerning this.

**Comment by rreusser** [comment]
Here's a clearer illustration of the coordinates:

The default camera defines zero. The scale of the units is defined by the window height, in pixels. A smaller fov camera operates on the exact same units as the default camera, so that the range could become slightly negative if you wanted fog much closer on a small fov camera. The main effect though is that fov does not affect the location of fog on terrain, which seems to me like the biggest concern.

cc @ansis @karimnaaji @arindam1993 


5e9331d9d4c51442bf8b10d01390b30c305b94f841f7b3987f705c6b4a75898f

**Comment by rreusser** [comment]
Ah, one question here is: does this mean we need to allow negative numbers in the fog range style spec? Does that affect slider values in studio?

**Comment by rreusser** [comment]
One other option would be to *simplify* the computation a bit by removing the reference offset so that this function:
```js
export function getFovAdjustedFogRange(state: FogState, fov: number): [number, number] {
    const shift = 100 * 0.5 / Math.tan(fov * 0.5) - FOG_REFERENCE_FOV_SHIFT;
    return [state.range[0] + shift, state.range[1] + shift];
}
```
simplifies to:
```js
export function getFovAdjustedFogRange(state: FogState, fov: number): [number, number] {
    const shift = 100 * 0.5 / Math.tan(fov * 0.5);
    return [state.range[0] + shift, state.range[1] + shift];
}
```

The downside then is that fog values may be slightly negative in many use cases (we already would have to allow negative values for narrow fov), but the upshot is that it's easier to understand the meaning. In rough terms, [-10, 10] would place the fog layer from 10% of the screen height in front of the center point to 10% of the screen height behind it.

Screen Shot 2021-04-30 at 3 55 48 PM copy

**Review by karimnaaji — APPROVED** [review]

**Review by karimnaaji — APPROVED** [review]
Looks good to me, switching to use the center as mentioned https://github.com/mapbox/mapbox-gl-js/pull/10629#issuecomment-830445582 is simpler to reason about compared to using our current FOV as a reference anchor (and introducing negative values for other FOVs).

File: src/style-spec/reference/v8.json [hunk 2]
```diff
@@ -3697,7 +3697,7 @@
           "zoom"
         ]
       },
-      "doc": "The start and end distance range in which fog fades from fully transparent to fully opaque. Units are relative to the map window height in CSS `vh` units, so that 200 corresponds to 200% the map window height.",
+      "doc": "The start and end distance range in which fog fades from fully transparent to fully opaque. The distance to the point at the center of the map is defined as zero, so that negative range values are closer to the camera, and positive values are farther away. Units are relative to the map window height in CSS `vh` units, so that 200 corresponds to 200% the map window height.",
```
[the remark this answers is not in the document]
Comment (rreusser) [reply src/style-spec/reference/v8.json:3700]: Hmm… maybe repetitive since it's farther away from both?

**Pull request merged**